kirin/queue_workers/__init__.py

```diff
@@ -119,108 +131,113 @@ def get_consumers(self, Consumer, channel):
-    if len(contributors) > 1:
-        logger.warning(
-            "more than one PIV contributors: {0}; choosing '{1}'".format(
-                map(lambda c: c.id, contributors), contributors[0].id
-            )
         )
-    return contributors[0] if len(contributors) > 0 else None
 
+    @abstractmethod
+    def get_connector_type(self):
+        raise NotImplementedError
 
-def try_db_commit():
-    try:
-        db.session.commit()
-    except Exception as db_e:
-        logger.warning("Exception while db-commit: {0}".format(db_e))
-        db.session.rollback()
 
+class QueueWorkerLauncher:
+    def __init__(self, logger: Logger, worker_class, connector_type):
+        self.logger = logger
+        self.worker_class = worker_class
+        self.connector_type = connector_type
+        self.lock_name = build_lock_name(self.connector_type)
 
-def attempt_piv_worker_launch():
-    with get_lock(logger, PIV_LOCK_NAME, PIV_WORKER_REDIS_TIMEOUT_LOCK.total_seconds()) as locked:
-        if locked:
-            should_wait = True
-            contributor = None
+    def process_loop(self):
+        while True:
             try:
-                contributor = pick_piv_contributor()
-                if not contributor:
-                    logger.warning("no PIV contributor")
-                    time.sleep(CONF_RELOAD_INTERVAL.total_seconds())
-                    return
-
-                with PivWorker(contributor) as worker:
-                    should_wait = False  # wait only after init crash
-                    logger.info("launching the PIV worker for '{0}'".format(contributor.id))
-                    worker.run()
-
-            except Exception as e:
-                logger.warning("PIV worker died: {0}".format(e))
-            finally:
-                try_db_commit()
-                if should_wait:
-                    time.sleep(CONF_RELOAD_INTERVAL.total_seconds())
-                if contributor is not None:
-                    db.session.expire(
-                        contributor
-                    )  # force db-reload otherwise staying locked on previous contributor's config
-        else:
-            logger.debug("Lock {%s} not acquired (already taken)", PIV_LOCK_NAME)
-            time.sleep(CONF_RELOAD_INTERVAL.total_seconds())
-
-
-@app.cli.command("piv_worker")
-def piv_worker():
-    # We assume a maximum of one PIV contributor is going to exist in the DB
-    # Therefore, we enforce a maximum of one `PivWorker` is running at a time
-    while True:
+                self.attempt_worker_launch()
+            except LockError as lock_e:
+                self.logger.warning("Exception with lock : {0}".format(lock_e))
+
+    def process(self) -> bool:
+        contributors = Contributor.find_by_connector_type(self.connector_type.value)
+        contributor = contributors[0] if len(contributors) > 0 else None
+        if not contributor:
+            raise NoContributorException
```

Comment:
You can probably avoid the raise here (and the try/except/return wrapping it):
```suggestion
            self.logger.warning("no contributor for {}".format(self.connector_type.value))
            return True  # Wait before next check that contributor exists
```